File: class22/python/GQS2201-CalculadoraMetricas.py
#!/usr/bin/env python3
"""
GQS2201 - Calculadora de Métricas de Software.
Calcula MTBF, MTTR, Disponibilidade e Densidade de Defeitos.
Uso: python3 GQS2201-CalculadoraMetricas.py
"""


def calcular_mtbf(tempo_total_h: float, num_falhas: int) -> float:
    if num_falhas == 0:
        return float("inf")
    return round(tempo_total_h / num_falhas, 2)


def calcular_mttr(tempo_reparo_h: float, num_falhas: int) -> float:
    if num_falhas == 0:
        return 0.0
    return round(tempo_reparo_h / num_falhas, 2)


def calcular_disponibilidade(mtbf: float, mttr: float) -> float:
    if mtbf + mttr == 0:
        return 0.0
    return round(mtbf / (mtbf + mttr) * 100, 4)


def calcular_densidade_defeitos(defeitos: int, loc: int) -> float:
    if loc == 0:
        return 0.0
    return round(defeitos / (loc / 1000), 2)


def nivel_disponibilidade(disponibilidade: float) -> str:
    if disponibilidade >= 99.999:
        return "Five Nines (Critical Systems)"
    elif disponibilidade >= 99.99:
        return "Four Nines (High Availability)"
    elif disponibilidade >= 99.9:
        return "Three Nines (Standard)"
    elif disponibilidade >= 99:
        return "Two Nines (Basic)"
    else:
        return "Below Standard"


def main():
    print("=" * 60)
    print("  GQS2201 - CALCULADORA DE MÉTRICAS DE SOFTWARE")
    print("=" * 60)

    print("\n--- Métricas de Confiabilidade ---")
    tempo_total = float(input("Tempo total de operação (horas): ") or 720)
    num_falhas = int(input("Número de falhas: ") or 3)
    tempo_reparo = float(input("Tempo total de reparo (horas): ") or 6)

    mtbf = calcular_mtbf(tempo_total, num_falhas)
    mttr = calcular_mttr(tempo_reparo, num_falhas)
    disp = calcular_disponibilidade(mtbf, mttr)

    print(f"\n  MTBF = {mtbf}h")
    print(f"  MTTR = {mttr}h")
    print(f"  Disponibilidade = {disp}%")
    print(f"  Nível: {nivel_disponibilidade(disp)}")

    print("\n--- Densidade de Defeitos ---")
    defeitos = int(input("Número de defeitos encontrados: ") or 25)
    loc = int(input("Tamanho do sistema (LOC): ") or 50000)
    densidade = calcular_densidade_defeitos(defeitos, loc)
    print(f"\n  Densidade de Defeitos = {densidade} defeitos/KLOC")
    if densidade < 1:
        print("  Classificação: Excelente (sistemas críticos)")
    elif densidade < 5:
        print("  Classificação: Bom (sistemas corporativos)")
    else:
        print("  Classificação: Precisa melhorar")


if __name__ == "__main__":
    main()

File: class22/java/GQS2201-CalculadoraMetricas.java
/**
 * GQS2201 - Calculadora de Métricas de Software.
 * Calcula MTBF, MTTR, Disponibilidade e Densidade de Defeitos.
 *
 * Compilação e execução:
 *   javac GQS2201-CalculadoraMetricas.java
 *   java -cp . GQS2201_CalculadoraMetricas
 */
import java.util.*;

class GQS2201_CalculadoraMetricas {

    static double calcularMTBF(double tempoTotalH, int numFalhas) {
        if (numFalhas == 0) return Double.POSITIVE_INFINITY;
        return Math.round(tempoTotalH / numFalhas * 100.0) / 100.0;
    }

    static double calcularMTTR(double tempoReparoH, int numFalhas) {
        if (numFalhas == 0) return 0.0;
        return Math.round(tempoReparoH / numFalhas * 100.0) / 100.0;
    }

    static double calcularDisponibilidade(double mtbf, double mttr) {
        if (mtbf + mttr == 0) return 0.0;
        return Math.round(mtbf / (mtbf + mttr) * 100 * 10000.0) / 10000.0;
    }

    static double calcularDensidadeDefeitos(int defeitos, int loc) {
        if (loc == 0) return 0.0;
        return Math.round(defeitos / (loc / 1000.0) * 100.0) / 100.0;
    }

    static String nivelDisponibilidade(double disponibilidade) {
        if (disponibilidade >= 99.999) {
            return "Five Nines (Critical Systems)";
        } else if (disponibilidade >= 99.99) {
            return "Four Nines (High Availability)";
        } else if (disponibilidade >= 99.9) {
            return "Three Nines (Standard)";
        } else if (disponibilidade >= 99) {
            return "Two Nines (Basic)";
        } else {
            return "Below Standard";
        }
    }

    /** Lê uma linha do Scanner; retorna o valor padrão se a entrada for vazia. */
    static double lerDoubleOuPadrao(Scanner sc, String prompt, double padrao) {
        System.out.print(prompt);
        String linha = sc.nextLine().strip();
        return linha.isEmpty() ? padrao : Double.parseDouble(linha);
    }

    static int lerIntOuPadrao(Scanner sc, String prompt, int padrao) {
        System.out.print(prompt);
        String linha = sc.nextLine().strip();
        return linha.isEmpty() ? padrao : Integer.parseInt(linha);
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("  GQS2201 - CALCULADORA DE MÉTRICAS DE SOFTWARE");
        System.out.println("=".repeat(60));

        try (Scanner sc = new Scanner(System.in)) {
            System.out.println("\n--- Métricas de Confiabilidade ---");
            double tempoTotal = lerDoubleOuPadrao(sc, "Tempo total de operação (horas): ", 720);
            int numFalhas = lerIntOuPadrao(sc, "Número de falhas: ", 3);
            double tempoReparo = lerDoubleOuPadrao(sc, "Tempo total de reparo (horas): ", 6);

            double mtbf = calcularMTBF(tempoTotal, numFalhas);
            double mttr = calcularMTTR(tempoReparo, numFalhas);
            double disp = calcularDisponibilidade(mtbf, mttr);

            System.out.printf("%n  MTBF = %sh%n", mtbf);
            System.out.printf("  MTTR = %sh%n", mttr);
            System.out.printf("  Disponibilidade = %s%%%n", disp);
            System.out.printf("  Nível: %s%n", nivelDisponibilidade(disp));

            System.out.println("\n--- Densidade de Defeitos ---");
            int defeitos = lerIntOuPadrao(sc, "Número de defeitos encontrados: ", 25);
            int loc = lerIntOuPadrao(sc, "Tamanho do sistema (LOC): ", 50000);
            double densidade = calcularDensidadeDefeitos(defeitos, loc);
            System.out.printf("%n  Densidade de Defeitos = %s defeitos/KLOC%n", densidade);
            if (densidade < 1) {
                System.out.println("  Classificação: Excelente (sistemas críticos)");
            } else if (densidade < 5) {
                System.out.println("  Classificação: Bom (sistemas corporativos)");
            } else {
                System.out.println("  Classificação: Precisa melhorar");
            }
        }
    }
}
